### ISSUE TITLE
DynamoDB call not record after enabling graphql dataloader

### ISSUE THREAD

**Issue by Blankll** [newrelic/newrelic-node-apollo-server-plugin#215]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
After enabling [graphql dataloader](https://github.com/graphql/dataloader), the DynamoDB call Span is not captured in lambda's distributed trace.
## Expected Behavior
Trace should record DynamoDB call span even if it triggered in dataloader

## Your Environment
env: AWS-lambda
Runtime: node-16
NewRelic-AWS Lambda-Layers: `arn:aws:lambda:ap-southeast-2:451483290750:layer:NewRelicNodeJS16X:16`
dependencies:
```json
{
   "@aws-sdk/client-dynamodb": "^3.128.0",
    "@aws-sdk/lib-dynamodb": "^3.128.0",
     "apollo-datasource": "^3.3.2",
    "apollo-server-lambda": "^3.10.1",
    "@newrelic/apollo-server-plugin": "^2.0.1"
}
```

**Comment by jordigh** [comment]
Hi, @Blankll thanks for the bug report. This seems related to other problems, and it would help if we can gather more information about what is happening.

For issues like this and other related issues, (https://github.com/newrelic/node-newrelic-aws-sdk/issues/145 https://github.com/newrelic/node-newrelic-aws-sdk/issues/146 and https://github.com/newrelic/node-newrelic-aws-sdk/issues/147), I recommend leveraging our support team over at support.newrelic.com which can help deep-dive into solving the problem. They will also ask to collect the necessary information which may or may not be sensitive (such as logs).

**Comment by Blankll** [comment]
Hi @jordigh the other 3 issues are more about missing some information in Span, but this one is different, I think this one is because this plugin currently hasn't compatible with `dataloader` or `dataloader` should also need to instrument, it is not some information missing in Span, it does not record the DynamoDB call Span that triggered in dataloader at all

**Comment by Blankll** [comment]
Hi @jordigh @bizob2828 this [repo](https://github.com/Blankll/serverless-event) has the sample  to simulateissue #145, issue #146, and issue #i147, please have a check

**Comment by bizob2828** [comment]
Much appreciated @Blankll.  We will take a look internally and get back to you.

**Comment by bizob2828** [comment]
@Blankll I responded to all the issues you logged except for this.  Can you update your example app to use dataloader so I can check out what's going on around losing context with dynamo?

**Comment by olivierlacan** [comment]
I don't think this issue is specific to DynamoDB, I believe it's specific to Dataloader itself and the way it executes all datastore queries as a batch: https://github.com/graphql/dataloader#batching

With this plugin enabled, New Relic database traces misreport time spent loading data within dataloader batch loading functions by a fairly wide margin: 
<img width="1362" alt="image" src="https://user-images.githubusercontent.com/65950/213690577-274f919e-7182-4ea9-bdcf-5c50a042fa41.png">

I can confirm the 762ms total load time is accurate based on browser timings of this same GraphQL resolver query: 
<img width="521" alt="image" src="https://user-images.githubusercontent.com/65950/213691008-3b2a6ad1-9f55-4ce1-937a-9e49017dbc4e.png">

This also fails to list a `Postgres select` within each of the `resolve.ApolloServer` function calls which each correspond to a batch loader function in a Dataloader class.

Since Dataloader uses `process.nextTick` this plugin (or the newrelic package, maybe?) might need to account for that. See: https://github.com/graphql/dataloader/blob/588a8b6c6391aad042b369f10dc440c7e0458312/src/index.js#L232-L246

**Comment by bizob2828** [comment]
@olivierlacan if you can create a sample app that reproduces this problem we'd be able to triage the issue much easier.  Do you think that's possible?  I suspect it's an issue with async context propagation and perhaps adding new behavior for this dataloader

**Comment by bizob2828** [comment]
I removed the bug flag as we technically don't support the [graphql dataloader plugin](https://github.com/graphql/dataloader)
